x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java

```diff
@@ -0,0 +1,376 @@
+     */
+    // TODO: In theory, handleNullToken should not be called at all if any credentials is extracted
+    // pkg-private for tests
+    void handleNullToken(Authenticator.Context context, ActionListener<Authentication> listener) {
+        final Authentication authentication;
+        if (context.getFallbackUser() != null) {
+            // TODO: assert we really haven't extract any token
+            logger.trace(
+                "No valid credentials found in request [{}], using fallback [{}]",
+                context.getRequest(),
+                context.getFallbackUser().principal());
+            Authentication.RealmRef authenticatedBy = new Authentication.RealmRef("__fallback", "__fallback", nodeName);
+            authentication = new Authentication(context.getFallbackUser(),
+                authenticatedBy,
+                null,
+                Version.CURRENT,
+                Authentication.AuthenticationType.INTERNAL,
+                Collections.emptyMap());
+        } else if (shouldFallbackToAnonymous(context)) {
+            logger.trace(
+                "No valid credentials found in request [{}], using anonymous [{}]",
+                context.getRequest(),
+                anonymousUser.principal());
+            Authentication.RealmRef authenticatedBy = new Authentication.RealmRef("__anonymous", "__anonymous", nodeName);
+            authentication = new Authentication(anonymousUser,
+                authenticatedBy,
+                null,
+                Version.CURRENT,
+                Authentication.AuthenticationType.ANONYMOUS,
+                Collections.emptyMap());
+        } else {
+            authentication = null;
+        }
+
+        if (authentication != null) {
+            // TODO: we can also support run-as for fallback users if needed
+            // TODO: the authentication for fallback user is now serialised in the inner threadContext
+            //       instead of at the AuthenticationService level
+            writeAuthToContext(context, authentication, listener);
+        } else {
+            final ElasticsearchSecurityException ese = context.getRequest().anonymousAccessDenied();
+            if (false == context.getUnsuccessfulMessages().isEmpty()) {
+                logger.debug("Authenticating with null credentials is unsuccessful in request [{}]" +
+                    " after unsuccessful attempts of other credentials", context.getRequest());
+                final ElasticsearchSecurityException eseWithPreviousCredentials =
+                    new ElasticsearchSecurityException("unable to authenticate anonymously", ese.status(), ese.getCause());
+                ese.getHeaderKeys().forEach(k -> eseWithPreviousCredentials.addHeader(k, ese.getHeader(k)));
+                addMetadata(context, eseWithPreviousCredentials);
+                listener.onFailure(eseWithPreviousCredentials);
+            } else {
```

Comment:
This is a rough way to inject the information about unsuccessful credentials into the response. This is done to avoid touch the auditing part of code. It can be simplified (and likely can be deleted) if we tighten the logic of handling fallback and anonymous users.